### PR TITLE
Major refactor of RestClient

### PR DESCRIPTION
**Pull request by aaraney**
Main features:

- `RestClient`: a class which offers asynchronous performance via a convenient serial wrapper, sqlite database GET request caching, GET request backoff, batch GET requests, and flexible url argument encoding.

- `Url`: Treat urls like to `pathlib.Path`s. Supports appending paths with `/` and appending url query parameters with `+`.

- `ClientSession`: Extension of [`aiohttp_client_cache`](https://github.com/JWCook/aiohttp-client-cache) that adds exponential backoff.

- `Variadic`: Join list or tuple of objects on some delimiter. Useful when query parameters utilize a non-`key=value&key=value2` pattern.

- `Alias` and `AliasGroup`: Immutable utility classes that simplify common API library development patterns (i.e. separation of backend uris from front end interfaces).

Distantly related to:  #70, #50, #14

The main feature of interest is the `RestClient` and its `mget` method. Basically you pass in a list of `urls` and it using `aiohttp` to get all of the requests using async. 

## Additions

- `Url`: Treat urls like to `pathlib.Path`s. Supports appending paths with `/` and appending url query parameters with `+`.

- `ClientSession`: Extension of [`aiohttp_client_cache`](https://github.com/JWCook/aiohttp-client-cache) that adds exponential backoff.

- `Variadic`: Join list or tuple of objects on some delimiter. Useful when query parameters utilize a non-`key=value&key=value2` pattern.

## Removals

- 

## Changes

- `RestClient`: a class which offers asynchronous performance via a convenient serial wrapper, sqlite database GET request caching, GET request backoff, batch GET requests, and flexible url argument encoding.
- RestClient parameters changed (breaking).

## Notes

- This functionality will be ported to `nwis_client` and `gcp_client` in future PRs.

## Checklist

- [x] PR has an informative and human-readable title
- [x] PR is well outlined and documented. See [#12](https://github.com/jarq6c/evaluation_tools/pull/12) for an example
- [x] Changes are limited to a single goal (no scope creep)
- [x] Code can be automatically merged (no conflicts)
- [x] Code follows project standards (see [CONTRIBUTING.md](../CONTRIBUTING.md))
- [ ] Passes all existing automated tests
- [x] Any _change_ in functionality is tested
- [x] New functions are documented (with a description, list of inputs, and expected output) using [numpy docstring](https://numpydoc.readthedocs.io/en/latest/format.html) formatting
- [x] Placeholder code is flagged / future todos are captured in comments
- [x] Reviewers requested with the [Reviewers tool](https://help.github.com/articles/requesting-a-pull-request-review/) :arrow_right:

### CODE REVIEW

**Comment by aaraney** [comment]
After 2c5ece0 it looks like the only test that is failing for `_restclient` is `test_variadic_integration_with_url` on `python` 3.7. I was able to reproduce it locally, so ill get something pushed up soon that resolves it.

```
2021-05-12T19:37:25.9392043Z     def test_variadic_integration_with_url():
2021-05-12T19:37:25.9392891Z         validation = "http://www.google.com?key=a,b,c"
2021-05-12T19:37:25.9393714Z >       url = Url("http://www.google.com") + {"key": Variadic(["a", "b", "c"])}
2021-05-12T19:37:25.9394148Z 
2021-05-12T19:37:25.9394819Z python/_restclient/tests/test_urllib.py:127: 
2021-05-12T19:37:25.9395528Z _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
2021-05-12T19:37:25.9396392Z python/_restclient/hydrotools/_restclient/urllib.py:82: in __add__
2021-05-12T19:37:25.9397052Z     return self.add(b)
2021-05-12T19:37:25.9397831Z python/_restclient/hydrotools/_restclient/urllib.py:125: in add
2021-05-12T19:37:25.9398613Z     query = parse.urlencode(query, doseq=True)
2021-05-12T19:37:25.9399753Z /opt/hostedtoolcache/Python/3.7.10/x64/lib/python3.7/urllib/parse.py:944: in urlencode
2021-05-12T19:37:25.9400668Z     v = quote_via(v, safe, encoding, errors)
2021-05-12T19:37:25.9401372Z /opt/hostedtoolcache/Python/3.7.10/x64/lib/python3.7/urllib/parse.py:854: in quote_plus
2021-05-12T19:37:25.9402125Z     return quote(string, safe, encoding, errors)
2021-05-12T19:37:25.9402860Z /opt/hostedtoolcache/Python/3.7.10/x64/lib/python3.7/urllib/parse.py:843: in quote
2021-05-12T19:37:25.9403537Z     return quote_from_bytes(string, safe)
```

**Comment by aaraney** [comment]
All the failing tests are `nwis_client`'s.

**Comment by jarq6c** [comment]
> All the failing tests are `nwis_client`'s.

How should we proceed? Merge this after review and get `nwis_client` working on a future PR or fix it here?

**Comment by aaraney** [comment]
> How should we proceed? Merge this after review and get `nwis_client` working on a future PR or fix it here?

That was my plan. I started refactoring the `nwis_client` this morning.

File: python/_restclient/README.md
Comment (jarq6c) [inline python/_restclient/README.md:19]: I'm a bit confused by the use of `Variadic` here. I understand this implements a variadic process, but I don't think it's applicable to any variadic process. The code below just looks like a `string.join` with some extras. Does the method work on any `object`?

File: python/_restclient/hydrotools/_restclient/_restclient.py
Comment (jarq6c) [inline python/_restclient/hydrotools/_restclient/_restclient.py:1]: Did you mean to replace the heading docstring with a shebang for a non-executable file?

File: python/_restclient/hydrotools/_restclient/_restclient.py
Comment (jarq6c) [inline python/_restclient/hydrotools/_restclient/_restclient.py:23]: I've noticed a lot of subjective terms in the documentation that I'm not sure help describe the code's function. Is it possible to summarize the inputs and outputs to a method without using words like "adept" or "robust"? At least at this stage, I'm not sure these are claims we can back-up with data. Some subjective terms are self-evident, like "simple" or "convenient," especially with regard to client tools. But I wonder if we keep the subjective descriptions in the README and leave docstrings a bit more sterile.

File: python/_restclient/hydrotools/_restclient/_restclient.py
Comment (jarq6c) [inline python/_restclient/hydrotools/_restclient/_restclient.py:40]: 👍🏽

File: python/_restclient/hydrotools/_restclient/async_client.py
Comment (jarq6c) [inline python/_restclient/hydrotools/_restclient/async_client.py:34]: Can you elaborate on why this is necessary in the comments? I found this issue from 2019 concerning this:

https://github.com/aio-libs/aiohttp/issues/3185

File: python/_restclient/setup.py
Comment (jarq6c) [inline python/_restclient/setup.py:24]: Honestly, I think the old description was more helpful.

File: python/_restclient/tests/test_client.py
Comment (jarq6c) [inline python/_restclient/tests/test_client.py:1]: Are these shebangs really necessary?

**Review by jarq6c — CHANGES_REQUESTED** [review]
Honestly, most of it is Greek to me. I understand this a major refactor and a huge change in the backbone of how we retrieve data, so I'm willing to give it a try. I wish we could incorporate it more incrementally. In the future we may want to discuss a limit on the size of PRs.

I'm a bit concerned about the `nwis_client` tests failing since that's really our only proof of improvement. It'd be great if the API for `_restclient` could quickly move to a more stable format over the next few iterations so that in the future downstream tools won't be effected so dramatically.

Much of my feedback had to with documentation and semantic choices. My documentation practices are certainly not perfect. In general, I tend to treat docstrings as technical descriptions with little to no subjective or colloquial phrasing. I assume software developers want a quick one or two lines that quickly summarize the inputs and outputs to see if a method is what they need. I try to limit "marketing language" to the README and then only in the initial description.

File: python/_restclient/README.md
Comment (aaraney) [reply python/_restclient/README.md:19]: It will work for any object that can be represented as a string. Admittedly the name might not be the most appropriate, do you have suggestions for an alternative name?

File: python/_restclient/hydrotools/_restclient/_restclient.py
Comment (aaraney) [reply python/_restclient/hydrotools/_restclient/_restclient.py:1]: Yes/No, i have my IDE automatically do that in python files for better or worse. Ill remove it and others that are not executable. Something I enabled a long time ago and never looked back.

File: python/_restclient/hydrotools/_restclient/_restclient.py
Comment (aaraney) [reply python/_restclient/hydrotools/_restclient/_restclient.py:23]: Certainly and thanks for the input. I agree that a more relaxed lexicon is more suitable for documentation.

File: python/_restclient/hydrotools/_restclient/async_client.py
Comment (aaraney) [reply python/_restclient/hydrotools/_restclient/async_client.py:34]: Why the comment is necessary? I just wanted to add context to why I ignored the warning. This is the same approach that `aiohttp_client_cache` takes to implement their `CachedSession` object. Personally I think throwing the deprecated warning is a bit overzealous. Of course, software will change and that will affect downstream users, but that's just part of the gig. 

My reasoning for filtering the warning is that the way ive implemented the class, it will always throw a warning to the user however it will not break usability.

File: python/_restclient/setup.py
Comment (aaraney) [reply python/_restclient/setup.py:24]: Okay, I may just drop "Abstract" as it can have undesired connotations. But I am happy to roll it back.

File: python/_restclient/tests/test_client.py
Comment (aaraney) [reply python/_restclient/tests/test_client.py:1]: Nope. Ill remove them. Nice catch.

File: python/_restclient/hydrotools/_restclient/async_client.py
Comment (jarq6c) [reply python/_restclient/hydrotools/_restclient/async_client.py:32]: I meant add why you are ignoring the warning as a comment.

Note: `aiohttp_client_cache` also minimally elaborate on why they ignore the warning and the implications.
https://github.com/JWCook/aiohttp-client-cache/blob/main/aiohttp_client_cache/session.py#L60

Example:
```python
# Ignore aiohttp.ClientSession Inheritance warning
#  This is the same approach that aiohttp_client_cache takes to implement their CachedSession object
```

Comment (aaraney) [reply python/_restclient/hydrotools/_restclient/async_client.py:32]: Ah, sorry I understand. Yeah Ill add some context in a comment.

File: python/_restclient/setup.py
Comment (jarq6c) [reply python/_restclient/setup.py]: I wasn't really advocating for a "roll back." What's nice about the old description is that it's more explicit about the capabilities offered by the package while also being succinct.

**Comment by jarq6c** [comment]
Note: #80 has been merged. So you may want to run your local tests against a Python 3.9 environment.

**Comment by jarq6c** [comment]
I'd say go ahead and feel free to merge into the alt branch, then we can work on getting the new backend working with `nwis_client` and `gcp_client` before merging into `main`.

**Comment by aaraney** [comment]
Agreed @jarq6c. Just for history sake. Offline, @jarq6c and I decided it would be best to create a new branch `restclient_transition` based off of `main` and add the changes here to that new branch to avoid breaking tests in the main branch. Eventually, `restclient_transition` will be rebased on `main`, but for now it is being treated as a dev branch. 

To develop off of `restclient_transition` it would be best to treat it like `main`. So, update your fork and create a new branch based off of `restclient_transition` then open a PR to merge changes back into `restclient_transition`.

**Comment by aaraney** [comment]
@hellkite500, when you get a chance could you please give your review of the code added in this PR?

**Pull request merged**
